### PR TITLE
lib: xlat_tables_v2: reduce time required to add a mmap region

### PR DESCRIPTION
**Pull request by vwadekar**
The last entry in the mapping table is not necessarily the same as the
end of the table. This patch loops through the table to find the last
entry marker, on every new mmap addition. The memove operation then
has to only move the memory between current entry and the last entry.
For platforms that arrange their MMIO map properly, this opearation
turns out to be a NOP.

The previous implementation added significant overhead per mmap
addition as the memmove operation always moved the difference between
the current mmap entry and the end of the table.

Tested on Tegra platforms and this new approach improves the memory
mapping time by ~75%, thus significantly reducing boot time on some
platforms.

Change-Id: Ie3478fa5942379282ef58bee2085da799137e2ca
Signed-off-by: Varun Wadekar <vwadekar@nvidia.com>

### CODE REVIEW

File: lib/xlat_tables_v2/xlat_tables_internal.c [hunk 1]
```diff
@@ -786,6 +786,19 @@ void mmap_add_region_ctx(xlat_ctx_t *ctx, const mmap_region_t *mm)
...
 
+	/*
+	 * Find the last entry marker in the mmap
+	 */
+	mm_last = ctx->mmap;
+	while ((mm_last->size != 0U) && (mm_last < mm_end)) {
+		++mm_last;
+	}
+
+	if (mm_last->size != 0U) {
```
Comment (antonio-nino-diaz-arm) [inline lib/xlat_tables_v2/xlat_tables_internal.c]: Actually, the function `mmap_add_region_check` checks that there is enough space in the mmap array. This should be an assert.

**Comment by vwadekar** [comment]
@antonio-nino-diaz-arm 

Thanks. I updated the patch to assert instead

**Comment by vwadekar** [comment]
@danh-arm @davidcunado-arm 

Does this change look good? Any comments?

**Comment by soby-mathew** [comment]
jenkins: test this please

**Comment by antonio-nino-diaz-arm** [comment]
The check that fails is the following one:

File | Function | Line number | Checker | Description
-- | -- | -- | -- | --
lib/xlat_tables_v2/xlat_tables_internal.c | mmap_add_region_ctx | 738 | MISRA C-2012 Rule 8.13 | [Advisory]  MISRA C-2012 Rule 8.13 violation: Consider adding "const" qualifier to  the pointer "mm_end" as destination object does not appear to be  modified within this function.

**Comment by vwadekar** [comment]
@antonio-nino-diaz-arm 

The latest patch fixes the MISRA warning

**Comment by dp-arm** [comment]
jenkins: test this please

**Pull request merged**